### ISSUE TITLE
No numbers in TypeVar names allowed

### ISSUE THREAD

**Issue by tyralla** [PyCQA/pylint#7428]
### Bug description

Since version 2.15, pylint reports the names of the following type variables to be invalid:

```python
from typing import TypeVar

T1 = TypeVar("T1") 
Float1 = TypeVar("Float1", bound=float) 
```

### Command used

```shell
pylint test_script.py
```

### Pylint output

```shell
C0103: Type variable name "T1" doesn't conform to predefined naming style (invalid-name)
 C0103: Type variable name "Float1" doesn't conform to predefined naming style (invalid-name)
```

### Expected behavior

No complaints about the TypeVar names.

### Pylint version

```shell
pylint 2.15.0
astroid 2.12.8
python 3.8.10
```

### OS / Environment

Windows (my computer), Linux (Travis-CI)

**Comment by DanielNoord** [comment]
@cdce8p Pinging you for opinion.

Shall I update the regular expression?

**Comment by cdce8p** [comment]
Cross posting here for visibility from https://github.com/PyCQA/pylint/pull/7322#issuecomment-1272340928

> The new regex also disallows things like `_T1` `_T2` which is a quite common practice

It's not as common as you may think. I would even argue that it's an antipattern which should be avoided except for very specific cases. It's much better to use different / more verbose names to avoid confusion and improve readability. Some common once include `S`, `T`, `U`, `V`, `KT`, `VT`. Just as an example take a look at the TypeVars defined in [`typeshed/stdlib/typing.pyi`](https://github.com/python/typeshed/blob/d1375f600cc4a7e7d81dcd9229c6abb933efda2a/stdlib/typing.pyi#L243-L252).

--
> Shall I update the regular expression?

I would leave it as is. Maybe we can update the documentation instead to suggest using alternative names?

**Comment by tyralla** [comment]
Thanks for keeping me informed!

Verbosity does not seem to be your main concern when favouring `S` and `T` over `T1` and `T2`.  Maybe it's only my impression that the last of the following definitions is the clearest, but at least it does not read that worse to call it an antipattern: 

```python
def f(x: S, y: T) -> tuple[T, S]: ...
def f(x: U, y: V) -> tuple[V, U]: ...
def f(x: T1, y: T2) -> tuple[T2, T1]: ...
```

So, my feeling is that pylint sets constraints here without strong (or at least transperant) reasons.  

Of course, switching to `S` and `T` in my code or changing a regex is also no big deal...

**Comment by Pierre-Sassoulas** [comment]
There wasn't a great deal of standard when we added this check so we took some decisions, but I don't see them as irreversible personally. In particular if this is formalized in a PEP we're going to change them.

**Comment by cdce8p** [comment]
> Thanks for keeping me informed!
> 
> Verbosity does not seem to be your main concern when favouring `S` and `T` over `T1` and `T2`. Maybe it's only my impression that the last of the following definitions is the clearest, but at least it does not read that worse to call it an antipattern:
> 
> ```python
> def f(x: S, y: T) -> tuple[T, S]: ...
> def f(x: U, y: V) -> tuple[V, U]: ...
> def f(x: T1, y: T2) -> tuple[T2, T1]: ...
> ```
> 
> So, my feeling is that pylint sets constraints here without strong (or at least transperant) reasons.
> 
> Of course, switching to `S` and `T` in my code or changing a regex is also no big deal...

Antipattern might have been a bit too strong of a word. Let's just say, I wouldn't recommend it personally. Looking at your example, IMO, with just number suffixes it's more likely they get confused than using different letters or even more descriptive names. In the end, I think readability is much more important than brevity.

The only good use of a number suffix I can think of is annotating some kind of `isinstance` method. Here is an example from astroid. _Although to be fair, it's more a workaround since the Python type system doesn't provide better options to annotate such a function. Maybe `Unpack` could be used in the future though._
https://github.com/PyCQA/astroid/blob/v2.12.10/astroid/nodes/node_ng.py#L53-L55
https://github.com/PyCQA/astroid/blob/v2.12.10/astroid/nodes/node_ng.py#L511-L561
Or another example for the builtins `map` function: [`typeshed/builtins.pyi`](https://github.com/python/typeshed/blob/3bcb0508f6324bbf3c08c33316b6fef036388297/stdlib/builtins.pyi#L1351-L1390). Similar case here that `_T1`, ..., `_T5` are used more as a workaround than anything else.

To suggest using better names, adding a simple `pylint: disable=invalid-name` in these cases is an acceptable tradeoff. Like Pierre said though, the default regex pattern isn't set in stone and can be adjusted in the future.

**Comment by tyralla** [comment]
Okay, thank you very much for elaborating.  Not my favourite decision, but also no problem.  Good to know, one can change the relevant regex pattern (I wasn't aware of this option).  Closing this issue as reporting not a bug but a new feature.
